nni/__main__.py

```diff
@@ -29,82 +29,50 @@ def main():
     exp_params = json.loads(exp_params_decode)
     logger.debug('exp_params json obj: [%s]', json.dumps(exp_params, indent=4))
 
-    if exp_params.get('deprecated', {}).get('multiThread'):
-        enable_multi_thread()
-
     if 'trainingServicePlatform' in exp_params:  # config schema is v1
         from .experiment.config.convert import convert_algo
-        for algo_type in ['tuner', 'assessor', 'advisor']:
+        for algo_type in ['tuner', 'assessor']:
             if algo_type in exp_params:
                 exp_params[algo_type] = convert_algo(algo_type, exp_params[algo_type])
+        if 'advisor' in exp_params:
+            exp_params['tuner'] = convert_algo('advisor', exp_params['advisor'])
 
-    if exp_params.get('advisor') is not None:
-        # advisor is enabled and starts to run
-        _run_advisor(exp_params)
-    else:
-        # tuner (and assessor) is enabled and starts to run
-        assert exp_params.get('tuner') is not None
-        tuner = _create_tuner(exp_params)
+    assert exp_params.get('tuner') is not None
+    tuner = _create_algo(exp_params['tuner'], 'tuner')
+
+    if isinstance(tuner, MsgDispatcherBase):  # is advisor
+        logger.debug(f'Tuner {type(tuner).__name__} is advisor.')
         if exp_params.get('assessor') is not None:
-            assessor = _create_assessor(exp_params)
-        else:
-            assessor = None
-        dispatcher = MsgDispatcher(tuner, assessor)
-
-        try:
-            dispatcher.run()
-            tuner._on_exit()
-            if assessor is not None:
-                assessor._on_exit()
-        except Exception as exception:
-            logger.exception(exception)
-            tuner._on_error()
-            if assessor is not None:
-                assessor._on_error()
-            raise
-
-
-def _run_advisor(exp_params):
-    if exp_params.get('advisor').get('name'):
-        dispatcher = create_builtin_class_instance(
-            exp_params['advisor']['name'],
-            exp_params['advisor'].get('classArgs'),
-            'advisors')
+            logger.error('Tuner {type(tuner).__name__} does not support early stopping. Assessor is ignored.')
+        tuner.run()
```

Comment:
just curious about the different between `tuner` and `advisor`, why can't adviser be a tuner?

Reply:
The root cause is that `Tuner` API is not powerful enough (because it designed for simplicity) and therefore some algorithms have to use lower level APIs. Tuning algorithms which uses low level APIs are (formerly) called "advisor".